src/histogram.c

```diff
@@ -0,0 +1,192 @@
+ * Usage:
+ *	 SELECT grouping_element, histogram(field, min, max, nbuckets) FROM table GROUP BY grouping_element.
+ *
+ * Description:
+ * Histogram generates a histogram array based off of a specified range passed into the function.
+ * Values falling outside of this range are bucketed into the 0 or nbucket+1 buckets depending on 
+ * if they are below or above the range, respectively. The resultant hitogram therefore contains
+ * nbucket+2 buckets accounting for buckets outside the range. 
+ */
+
+PG_FUNCTION_INFO_V1(hist_sfunc);
+PG_FUNCTION_INFO_V1(hist_combinefunc);
+PG_FUNCTION_INFO_V1(hist_serializefunc);
+PG_FUNCTION_INFO_V1(hist_deserializefunc);
+PG_FUNCTION_INFO_V1(hist_finalfunc);
+
+/* histogram(state, val, min, max, nbuckets) */
+Datum
+hist_sfunc(PG_FUNCTION_ARGS) 
+{
+	MemoryContext aggcontext; 
+	bytea 	*state = (PG_ARGISNULL(0) ? NULL : PG_GETARG_BYTEA_P(0));
+	Datum   *hist;
+
+	float 	val = PG_GETARG_FLOAT4(1); 
+	float 	min = PG_GETARG_FLOAT4(2); 
+	float 	max = PG_GETARG_FLOAT4(3); 
+	int 	nbuckets = PG_GETARG_INT32(4); 
+
+	int 	bucket = DirectFunctionCall4(width_bucket_float8, val, min, max, nbuckets);  
+
+	if (!AggCheckCallContext(fcinfo, &aggcontext)) {
+		/* cannot be called directly because of internal-type argument */
+		elog(ERROR, "hist_sfunc called in non-aggregate context");
+	}
+
+	if (min > max) {
+		/* cannot generate a histogram with incompatible bounds */
+		elog(ERROR, "lower bound cannot exceed upper bound");
+	}
+
+	if (state == NULL)
+    {
+    	/* Allocate memory to a new histogram state array */
+		Size arrsize = sizeof(Datum) * (nbuckets + 2);
+		state = MemoryContextAllocZero(aggcontext, VARHDRSZ + arrsize);
+		SET_VARSIZE(state, VARHDRSZ + arrsize);
+    }
+
+    /* Increment the proper histogram bucket */
+    hist = (Datum *) VARDATA(state);
+    hist[bucket] = Int32GetDatum(DatumGetInt32(hist[bucket]) + 1);
+
+	PG_RETURN_BYTEA_P(state);
+}
+
+/* Make a copy of the bytea state */
+static inline Datum
+copy_state(MemoryContext aggcontext, bytea *state) {
+	bytea *copy;
+	Size i;
+	Size arrsize = VARSIZE(state) - VARHDRSZ;
+	Datum *hist, *hist_copy;
+
+	copy = MemoryContextAllocZero(aggcontext, VARHDRSZ + arrsize);
+	SET_VARSIZE(copy, VARHDRSZ + arrsize);
+
+	hist = (Datum *) VARDATA(state);
+	hist_copy = (Datum *) VARDATA(copy);
+
+	for (i = 0; i < (arrsize / sizeof(Datum)); i++)
+		hist_copy[i] = hist[i];
+
+	PG_RETURN_BYTEA_P(copy);
+}
+
+/* hist_combinefunc(internal, internal) => internal */
+Datum
+hist_combinefunc(PG_FUNCTION_ARGS)
+{
+	MemoryContext 	aggcontext;
+
+	bytea 	*state1 = (PG_ARGISNULL(0) ? NULL : PG_GETARG_BYTEA_P(0));
+	bytea 	*state2 = (PG_ARGISNULL(1) ? NULL : PG_GETARG_BYTEA_P(1));
+	bytea 	*result;
+
+    if (!AggCheckCallContext(fcinfo, &aggcontext))
+	{
+		/* cannot be called directly because of internal-type argument */
+		elog(ERROR, "hist_combinefunc called in non-aggregate context");
+	}
+
+	if (state2 == NULL) {
+		result = (bytea *) copy_state(aggcontext, state1);
```

Comment:
cast goes away here if you return `bytea *` from `copy_state()` as suggested above.